image_transport/src/republish_nodelet.cpp

```diff
@@ -0,0 +1,116 @@
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/o2r other materials provided
+ *     with the distribution.
+ *   * Neither the name of the JSK Lab nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+/*
+ * Author: Yuki Furuta <furushchev@jsk.imi.i.u-tokyo.ac.jp>
+ */
+
+
+#include <image_transport/image_transport.h>
+#include <image_transport/nodelet_lazy.h>
+#include <image_transport/publisher_plugin.h>
+#include <pluginlib/class_loader.h>
+
+
+namespace image_transport
+{
+class RepublishNodelet : public NodeletLazy
+{
+protected:
+  boost::shared_ptr<ImageTransport> it_;
+  Publisher pub_;
+  boost::shared_ptr<pluginlib::ClassLoader<PublisherPlugin> > loader_;
+  boost::shared_ptr<PublisherPlugin> pub_plugin_;
+  Subscriber sub_;
+  std::string in_transport_, out_transport_;
+
+  virtual void onInit()
+  {
+    NodeletLazy::onInit();
+
+    pnh_->getParam("in_transport", in_transport_);
+    pnh_->param<std::string>("out_transport", out_transport_, std::string());
+
+    it_.reset(new ImageTransport(*nh_));
+
+    if (out_transport_.empty())
+    {
+      pub_ = advertiseImage(*nh_, "out", 1);
+    }
+    else
+    {
+      loader_.reset(new pluginlib::ClassLoader<PublisherPlugin>(
+          "image_transport", "image_transport::PublisherPlugin"));
+      std::string lookup_name = PublisherPlugin::getLookupName(out_transport_);
+      pub_plugin_ = loader_->createInstance(lookup_name);
+
+      advertiseImage(*nh_, "out", 1, boost::weak_ptr<PublisherPlugin>(pub_plugin_));
+    }
+
+    onInitPostProcess();
+  }
+
+  virtual void subscribe()
+  {
+    std::string in_topic = nh_->resolveName("in");
+    if (out_transport_.empty())
+    {
+      // Use Publisher::publish as the subscriber callback
+      typedef void (Publisher::*PublishMemFn)(const sensor_msgs::ImageConstPtr&) const;
+      PublishMemFn pub_mem_fn = &Publisher::publish;
+      sub_ = it_->subscribe(
```

Comment:
This might yield a warning from the ReconfigureServer of the CompressedSubscriber. Since a new subscriber is created before the old one is destroyed, the new Server wants to advertise an existing service. My solution would be, to manually destroy the sub_ object in unsubscribe().

Reply:
@Sefrin Thanks for the review! I could reproduced the bug,

```bash
# Terminal 1
roscore &
rosrun uvc_camera uvc_camera_node &
rosrun image_transport republish in:=image_raw compressed out:=foo
```

```bash
# Terminal 2
rostopic hz /foo/compressed
^C
rostopic hz /foo/compressed
```

And I could see the error message on the terminal 1:

```bash
furushchev@dango:~/ros/kinetic_test/src/image_common/image_transport/src [republish-nodelet]$ rosrun image_transport republish in:=image_raw compressed out:=foo
[ INFO] [1533828252.220730375] [/image_republisher_1533828252119766128:ros.nodelet]: Initializing nodelet with 8 worker threads.
[ WARN] [1533828257.226403951] [/image_republisher_1533828252119766128:ros.image_transport./image_republisher_1533828252119766128]: This node/nodelet subscribes topics only when subscribed.
[ERROR] [1533828263.122396843] [/image_republisher_1533828252119766128:ros.roscpp]: Tried to advertise a service that is already advertised in this node [/image_republisher_1533828252119766128/compressed/set_parameters]
```